File: controller/routes/logs.py
"""Log and SSE streaming endpoints."""

from __future__ import annotations

import asyncio
import datetime as dt
from collections import deque
from pathlib import Path

from fastapi import APIRouter, Depends, HTTPException
from fastapi.responses import StreamingResponse

from ..config import settings
from ..events import event_manager, Event
from ..process import find_inference_process
from ..store import RecipeStore

router = APIRouter()


def get_store() -> RecipeStore:
    """Get recipe store instance."""
    from ..app import get_store as _get_store
    return _get_store()


def _log_path_for(session_id: str) -> Path:
    safe = "".join(ch for ch in (session_id or "") if ch.isalnum() or ch in ("-", "_", "."))
    if not safe:
        raise HTTPException(status_code=400, detail="Invalid log session id")
    return Path("/tmp") / f"vllm_{safe}.log"


def _tail_lines(path: Path, limit: int) -> list[str]:
    try:
        with path.open("r", encoding="utf-8", errors="replace") as f:
            return list(deque(f, maxlen=limit))
    except FileNotFoundError:
        raise HTTPException(status_code=404, detail="Log not found")


@router.get("/logs", tags=["Logs"])
async def list_logs(store: RecipeStore = Depends(get_store)):
    """List available inference log files."""
    current = find_inference_process(settings.inference_port)
    log_files = sorted(Path("/tmp").glob("vllm_*.log"), key=lambda p: p.stat().st_mtime, reverse=True)

    sessions = []
    for p in log_files:
        sid = p.name.removeprefix("vllm_").removesuffix(".log")
        recipe = store.get(sid)
        created_at = dt.datetime.fromtimestamp(p.stat().st_mtime, tz=dt.timezone.utc).isoformat()

        status = "stopped"
        if current and recipe and current.model_path and recipe.model_path and recipe.model_path in current.model_path:
            status = "running"
        elif current and recipe and current.served_model_name and recipe.served_model_name == current.served_model_name:
            status = "running"

        sessions.append(
            {
                "id": sid,
                "recipe_id": recipe.id if recipe else sid,
                "recipe_name": recipe.name if recipe else None,
                "model_path": recipe.model_path if recipe else None,
                "model": (recipe.served_model_name or recipe.name) if recipe else sid,
                "backend": recipe.backend.value if recipe else None,
                "created_at": created_at,
                "status": status,
            }
        )

    return {"sessions": sessions}


@router.get("/logs/{session_id}", tags=["Logs"])
async def get_logs(session_id: str, limit: int = 2000):
    """Get log content for a session."""
    limit = max(1, min(int(limit), 20000))
    path = _log_path_for(session_id)
    lines = _tail_lines(path, limit)
    logs = [ln.rstrip("\n") for ln in lines]
    return {"id": session_id, "logs": logs, "content": "\n".join(logs)}


@router.delete("/logs/{session_id}", tags=["Logs"])
async def delete_logs(session_id: str):
    """Delete a log file."""
    path = _log_path_for(session_id)
    try:
        path.unlink()
    except FileNotFoundError:
        raise HTTPException(status_code=404, detail="Log not found")
    return {"success": True}


@router.get("/events", tags=["Events"])
async def events_stream():
    """Subscribe to real-time status updates via Server-Sent Events."""
    async def event_generator():
        try:
            async for event in event_manager.subscribe():
                yield event.to_sse()
        except asyncio.CancelledError:
            pass

    return StreamingResponse(
        event_generator(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache, no-transform",
            "Connection": "keep-alive",
            "X-Accel-Buffering": "no",
        }
    )


@router.get("/logs/{session_id}/stream", tags=["Logs"])
async def stream_logs(session_id: str):
    """Stream log file updates in real-time via SSE."""
    path = _log_path_for(session_id)

    async def log_generator():
        try:
            if path.exists():
                with path.open("r", encoding="utf-8", errors="replace") as f:
                    for line in f:
                        event = Event(type="log", data={"line": line.rstrip("\n")})
                        yield event.to_sse()

            async for event in event_manager.subscribe(f"logs:{session_id}"):
                yield event.to_sse()
        except asyncio.CancelledError:
            pass

    return StreamingResponse(
        log_generator(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache, no-transform",
            "Connection": "keep-alive",
        }
    )


@router.get("/events/stats", tags=["Events"])
async def events_stats():
    """Get event manager statistics for monitoring."""
    return event_manager.get_stats()

File: controller-new/src/routes/logs.ts
import type { Hono } from "hono";
import { readFileSync, unlinkSync, existsSync, readdirSync, statSync } from "node:fs";
import { join } from "node:path";
import type { AppContext } from "../types/context";
import { badRequest, notFound } from "../core/errors";
import { streamAsyncStrings, buildSseHeaders } from "../http/sse";
import { Event } from "../services/event-manager";

/**
 * Register log and SSE routes.
 * @param app - Hono app.
 * @param context - App context.
 */
export const registerLogsRoutes = (app: Hono, context: AppContext): void => {
  /**
   * Resolve log file path for a session id.
   * @param sessionId - Session identifier.
   * @returns Path to log file.
   */
  const logPathFor = (sessionId: string): string => {
    const safe = Array.from(sessionId).filter((char) => /[a-zA-Z0-9._-]/.test(char)).join("");
    if (!safe) {
      throw badRequest("Invalid log session id");
    }
    return join("/tmp", `vllm_${safe}.log`);
  };

  /**
   * Tail the last N lines of a file.
   * @param path - Log file path.
   * @param limit - Line limit.
   * @returns Array of lines.
   */
  const tailLines = (path: string, limit: number): string[] => {
    if (!existsSync(path)) {
      throw notFound("Log not found");
    }
    const content = readFileSync(path, "utf-8");
    const lines = content.split("\n");
    return lines.slice(Math.max(0, lines.length - limit));
  };

  app.get("/logs", async (ctx) => {
    const current = await context.processManager.findInferenceProcess(context.config.inference_port);
    const entries = readdirSync("/tmp")
      .filter((name) => name.startsWith("vllm_") && name.endsWith(".log"))
      .map((name) => ({ name, mtime: statSync(join("/tmp", name)).mtimeMs }))
      .sort((left, right) => right.mtime - left.mtime)
      .map((entry) => entry.name);
    const sessions = [];
    for (const filename of entries) {
      const sessionId = filename.replace(/^vllm_/, "").replace(/\.log$/, "");
      const recipe = context.stores.recipeStore.get(sessionId);
      const stat = statSync(join("/tmp", filename));
      const modifiedAt = new Date(stat.mtimeMs).toISOString();
      let status = "stopped";
      if (current && recipe && current.model_path && recipe.model_path && current.model_path.includes(recipe.model_path)) {
        status = "running";
      } else if (current && recipe && current.served_model_name && recipe.served_model_name === current.served_model_name) {
        status = "running";
      }
      sessions.push({
        id: sessionId,
        recipe_id: recipe?.id ?? sessionId,
        recipe_name: recipe?.name ?? null,
        model_path: recipe?.model_path ?? null,
        model: recipe ? (recipe.served_model_name ?? recipe.name) : sessionId,
        backend: recipe?.backend ?? null,
        created_at: modifiedAt,
        status,
      });
    }
    return ctx.json({ sessions });
  });

  app.get("/logs/:sessionId", async (ctx) => {
    const sessionId = ctx.req.param("sessionId");
    const limit = Math.min(Math.max(Number(ctx.req.query("limit") ?? 2000), 1), 20000);
    const path = logPathFor(sessionId);
    const lines = tailLines(path, limit).map((line) => line.replace(/\n$/, ""));
    return ctx.json({ id: sessionId, logs: lines, content: lines.join("\n") });
  });

  app.delete("/logs/:sessionId", async (ctx) => {
    const sessionId = ctx.req.param("sessionId");
    const path = logPathFor(sessionId);
    try {
      unlinkSync(path);
    } catch {
      throw notFound("Log not found");
    }
    return ctx.json({ success: true });
  });

  app.get("/events", async (_ctx) => {
    const stream = streamAsyncStrings((async function* (): AsyncGenerator<string> {
      for await (const event of context.eventManager.subscribe()) {
        yield event.toSse();
      }
    })());
    return new Response(stream, {
      headers: buildSseHeaders(),
    });
  });

  app.get("/logs/:sessionId/stream", async (ctx) => {
    const sessionId = ctx.req.param("sessionId");
    const path = logPathFor(sessionId);
    const stream = streamAsyncStrings((async function* (): AsyncGenerator<string> {
      if (existsSync(path)) {
        const content = readFileSync(path, "utf-8");
        for (const line of content.split("\n")) {
          if (line.length === 0) {
            continue;
          }
          const event = new Event("log", { line });
          yield event.toSse();
        }
      }
      for await (const event of context.eventManager.subscribe(`logs:${sessionId}`)) {
        yield event.toSse();
      }
    })());

    return new Response(stream, {
      headers: buildSseHeaders({ "Cache-Control": "no-cache, no-transform", Connection: "keep-alive" }),
    });
  });

  app.get("/events/stats", async (ctx) => {
    return ctx.json(context.eventManager.getStats());
  });
};
